File: implement-shell-tools/cat/cat.js
const fs = require("fs");

const args = process.argv.slice(2);

const numberAll = args[0] === "-n";
const numberNonBlank = args[0] === "-b";

const files = (numberAll || numberNonBlank)
    ? args.slice(1)
    : args;

let lineNumber = 1;

for (const file of files) {
    try {
        const content = fs.readFileSync(file, "utf8");
        const lines = content.endsWith("\n")
            ? content.slice(0, -1).split("\n")
            : content.split("\n");

        for (const line of lines) {
            if (numberAll) {
                console.log(`${String(lineNumber).padStart(6)}\t${line}`);
                lineNumber++;
            } else if (numberNonBlank && line !== "") {
                console.log(`${String(lineNumber).padStart(6)}\t${line}`);
                lineNumber++;
            } else {
                console.log(line);
            }
        }
    } catch (err) {
        console.error(`cat: ${file}: ${err.message}`);
    }
}

File: implement-shell-tools/cat/cat.py
import sys

args = sys.argv[1:]

number_all = "-n" in args
number_nonblank = "-b" in args

files = [arg for arg in args if arg not in ("-n", "-b")]

line_number = 1

for filename in files:
    try:
        with open(filename, "r") as file:
            content = file.read()

            lines = content[:-1].split("\n") if content.endswith("\n") else content.split("\n")

            for line in lines:
                if number_all:
                    print(f"{line_number:6}\t{line}")
                    line_number += 1

                elif number_nonblank and line != "":
                    print(f"{line_number:6}\t{line}")
                    line_number += 1

                else:
                    print(line)

    except Exception as err:
        print(f"cat: {filename}: {err}")